fjord/suggest/providers/trigger/tests/test_models.py

```diff
@@ -0,0 +1,189 @@
+
+    def test_match_keywords(self):
+        tests = [
+            # tr.keywords, description, expected
+            (u'', u'', True),
+            (u'', u'my dog has fleas', True),
+            (u'my', u'', False),
+            (u'my', u'my dog has fleas', True),
+            (u'my', u'MY DOG HAS FLEAS', True),
+            (u'your\ndog', u'my dog has fleas', True),
+            (u'my dog', u'my dog has fleas', True),
+            (u'my dog', u'your dog has fleas', False),
+            (u'my dog\ncat', u'my dog has fleas', True),
+            (u'my cat', u'my dog has fleas', False),
+            (u'my cat', u'my  cat  hat  fleas', True),
+            (u'my  cat', u'my  cat  hat  fleas', True),
+            (u'\xca', u'abcd \xca abcd', True),
+            (u'my-cat', u'my-cat has fleas', True),
+            (u'(my cat)', u'(my cat) has fleas', True),
+        ]
+        for tr_keywords, description, expected in tests:
+            tr = TriggerRuleFactory(keywords=tr_keywords)
+            assert tr.match_description(description) == expected
+
+    def test_match_url_exists(self):
+        tests = [
+            # tr.url_exists, url, expected
+            (None, '', True),
+            (None, 'https://example.com/', True),
+            (True, '', False),
+            (True, 'https://example.com/', True),
+            (False, '', True),
+            (False, 'https://example.com/', False)
+        ]
+        for tr_url_exists, url, expected in tests:
+            tr = TriggerRuleFactory(url_exists=tr_url_exists)
+            assert tr.match_url_exists(url) == expected
+
+    def test_match(self):
+        # Note: This isn't an exhaustive test. Just a rough cursory check.
+
+        # TriggerRule that matches everything matches everything.
+        # Note: products=[] is implied.
+        tr = TriggerRuleFactory(
+            versions=u'',
+            locales=u'',
+            keywords=u''
+        )
+
+        resp = ResponseFactory()
+        assert tr.match(resp) is True
+
+        tr = TriggerRuleFactory(
+            versions=u'38*',
+            locales=u'en-US,fr',
+            keywords=u'rc4',
+            url_exists=True
+        )
+        prod = ProductFactory()
+        tr.products.add(prod)
+        resp = ResponseFactory(
+            version=u'38.0.5',
+            locale=u'en-US',
+            product=prod.db_name,
+            description=u'rc4 is awesome',
+            url=u'https://example.com/'
+        )
+        assert tr.match(resp) is True
+        resp.locale = 'es'
+        assert tr.match(resp) is False
+
+    def test_keyword_regexes_are_cached(self):
+        for text in [u'cat', u'dog', u'mouse']:
+            tr = TriggerRuleFactory(keywords=text)
+            tr.match_description(text)
+
+            tr = TriggerRuleFactory(keywords=text)
+            tr.match_description(text)
+
+        cache_info = _generate_keywords_regex.cache_info()
+        assert cache_info.hits >= 3
```

Comment:
I've never seen a test like this, though I can imagine why it is important that these regexes are cached. Did you run into this in development? Or just wanted to be very sure this would work well?